### DISCUSSION TITLE
Clarification of search configuration and queries

### DISCUSSION THREAD

**Discussion by tscahill** [zim-desktop-wiki/zim-desktop-wiki#2705] [Q&A]
So, until just now, I have had a faulty confidence in the search capability built into Zim. I assumed word stemming and alternate case were baked in, because my searches always seem to bring exactly what I was looking for. 

Yesterday, to better keep track of my regime of pills and supplements, I added a section "Current Prescriptions" to a page called 'Health'. Then I filled in all the details of each prescription number, etc.. 

Since then, I had been working heavily in other areas of information, and afterwards, I opened the search window and entered "prescription", and then "prescriptions", but neither query provided the desired result. When I changed the query to "Prescriptions", the page I was looking for appeared in the results list.

The "Find" dialog requires one to turn on 'Match case' to get that filter. The built-in help does not mention case sensitivity or word stemming as part of the "Search->Search...". 

So, is this a feature or a bug? I've had a presumption that I could type "charle" into the search dialog and the results would include Charles and Charlemagne. As one inclined to error, my propensity has been to reduce the impact of that by keeping my searches simple and truncated. This increases my chances of finding some specific text, even if I had spelled it wrong. 

Now I'm wondering how many times I've thought I'd done a comprehensive search, but had actually missed the plot.

**Comment by maclomhair** [comment]
My understanding is that by default, the search (from the search window) should be case-insensitive all the time, which would mean that any matches that only work if written in the correct case are bugs.

Also, my understanding is that searching by default only includes entire word matches, so searching for "charle" should yield neither "Charles" nor "Charlemagne"... but the `*` wildcard is supported, so searching for "charle*" should do the trick.

I hope this helps!

**Reply by tscahill** [reply]
Thanks for chipping in! I was quite surprised that upper and lower case queries would be determinant. I'd like to understand this better. I've been perusing this project, looking for insight to search functionality. I assume the capability comes imported from some somewhere else in the python ecology. 

Is this the relevant code?
https://github.com/zim-desktop-wiki/zim-desktop-wiki/blob/develop/zim/search.py

Line 43 seems to imply that the search is for an input "string"

My presence is now required for a Costco run. I'll pick up reviewing that code when the perishables are stowed. If I'm on the wrong track, please kindly let me know... : )

**Reply by maclomhair** [reply]
Yes, I think that's the relevant code, and for full-text search in particular [line 558 and following](https://github.com/zim-desktop-wiki/zim-desktop-wiki/blob/develop/zim/search.py#L558)! 

As far as I understand the code, for each search term, a regular expression is first constructed (line 573, which calls the `_content_regex` method). This regular expression is where case-insensitivity comes in: `_content_regex` takes an optional case sensitivity parameter which defaults to false, so the regex is built with pythons `re.I` option, indicating case insensitive matching.

This regex is then matched against all text in each page, see [line 608](https://github.com/zim-desktop-wiki/zim-desktop-wiki/blob/develop/zim/search.py#L608) for example.

Beware, though, that due to the way the page parsing works, the search misses words which are *partially* styled, like "Cha**le**s" - for the search functionality, this will look like three different words: "Cha", "le" and "s".

**Comment by jaap-karssenberg** [comment]
I agree the documentation is not clear and the search dialoog lacks visual
feedback on this behavior.

The default is not case sensitive, but it does do whole word matches only
unless you use the "*" syntax.

Will think about how to improve this.

Op vr 13 dec. 2024 12:15 schreef maclomhair ***@***.***>:

**Comment by jaap-karssenberg** [comment]
Yes the search functionality is defined in "Zim/search.py".  Be aware that
I'm drafting ideas how to refactor this code. So if you plan to modify or
reuse make sure to not depend on the internals to much.

Regards,

Jaap

Op vr 13 dec. 2024 18:09 schreef maclomhair ***@***.***>:
